core/src/ledger_cleanup_service.rs

```diff
@@ -100,43 +100,61 @@ impl LedgerCleanupService {
+            .iter()
+            .filter(|live_file| live_file.column_family_name == data_shred_cf_name)
+            .map(|file_meta| file_meta.num_entries)
+            .sum();
+
+        if num_shreds <= max_ledger_shreds {
+            return (false, 0, num_shreds);
         }
-        iterate_time.stop();
+
+        // Using the difference between the lowest and highest slot seen will
+        // result in overestimating the number of slots in the blockstore since
+        // there are likely to be some missing slots, such as when a leader is
+        // delinquent for their leader slots.
+        //
+        // With the below calculations, we will then end up underestimating the
+        // mean number of shreds per slot present in the blockstore which will
+        // result in cleaning more slots than necessary to get us
+        // below max_ledger_shreds.
+        //
+        // Given that the service runs on an interval, this is good because it
+        // means that we are building some headroom so the peak number of alive
+        // shreds doesn't get too large before the service's next run.
+        let lowest_slot = blockstore.lowest_slot();
+        let highest_slot = blockstore
+            .highest_slot()
+            .expect("Blockstore::highest_slot()")
+            .unwrap_or(lowest_slot);
+        // The + 1 ensures we count the correct number of slots and that the
+        // divisor in the subsequent division is strictly >= 1.
+        let num_slots = highest_slot - lowest_slot + 1;
+        let mean_shreds_per_slot = num_shreds / num_slots;
         info!(
-            "total_slots={} total_shreds={} max_ledger_shreds={}, {}",
-            total_slots.len(),
-            total_shreds,
-            max_ledger_shreds,
-            iterate_time
+            "{} alive shreds in slots [{}, {}], mean of {} shreds per slot",
+            num_shreds, lowest_slot, highest_slot, mean_shreds_per_slot
         );
-        if total_shreds < max_ledger_shreds {
-            return (false, 0, total_shreds);
-        }
-        let mut num_shreds_to_clean = 0;
-        let mut lowest_cleanup_slot = total_slots[0].0;
-        for (slot, num_shreds) in total_slots.iter().rev() {
-            num_shreds_to_clean += *num_shreds;
-            if num_shreds_to_clean > max_ledger_shreds {
-                lowest_cleanup_slot = *slot;
-                break;
-            }
-        }
 
-        (true, lowest_cleanup_slot, total_shreds)
+        // Add an extra (mean_shreds_per_slot - 1) in the numerator
+        // so that our integer division rounds up
+        let num_slots_to_clean = (num_shreds - max_ledger_shreds + mean_shreds_per_slot - 1)
+            .checked_div(mean_shreds_per_slot);
+
+        if let Some(num_slots_to_clean) = num_slots_to_clean {
+            // Ensure we don't cleanup anything past the last root we saw
+            let lowest_cleanup_slot = std::cmp::min(lowest_slot + num_slots_to_clean - 1, root);
```

Comment:
This is great!  I am specifically looking for the root check in this PR, and I'm glad this is something that is already included!

Btw, I am thinking about whether we should do `root` or `root - 1`.  Is there any benefit that the ledger-store always keeps a root?

Reply:
Good question. From a functional standpoint, we would want to keep all shreds past the latest snapshot (noting that snapshot slots are always roots). If the node were to be restarted, we would need to replay all those slots since the snapshot during catchup.

Snapshots are taken fairly frequently, I think default on incremental is 100 slots (or that order of magnitude at least). With our current allowed min ledger size of 50M shreds and assuming 1k shreds per slot, that is 50k slots we keep around. We enter this loop whenever the latest root had advanced >= 512 slots since the last run, so we are never close to purging new data

That is, looking at `std::cmp::min(lowest_slot + num_slots_to_clean - 1, root);`, the min will always be the left hand side in practice.